lib/streamlit/runtime/context.py

```diff
@@ -0,0 +1,154 @@
+# Copyright (c) Streamlit Inc. (2018-2022) Snowflake Inc. (2022-2024)
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import annotations
+
+from functools import lru_cache
+from types import MappingProxyType
+from typing import TYPE_CHECKING, Any, Iterable, Iterator, Mapping, cast
+
+from streamlit import runtime
+from streamlit.runtime.metrics_util import gather_metrics
+from streamlit.runtime.scriptrunner import get_script_run_ctx
+from streamlit.type_util import is_type
+
+if TYPE_CHECKING:
+    from http.cookies import Morsel
+
+    from tornado.httputil import HTTPHeaders, HTTPServerRequest
+    from tornado.web import RequestHandler
+
+
+def _get_request() -> HTTPServerRequest | None:
+    ctx = get_script_run_ctx()
+    if ctx is None:
+        return None
+
+    session_client = runtime.get_instance().get_client(ctx.session_id)
+    if session_client is None:
+        return None
+
+    # We return websocket request only if session_client is an instance of
+    # BrowserWebSocketHandler (which is True for the Streamlit open-source
+    # implementation). For any other implementation, we return None.
+    if not is_type(
+        session_client,
+        "streamlit.web.server.browser_websocket_handler.BrowserWebSocketHandler",
+    ):
+        return None
+    return cast("RequestHandler", session_client).request
+
+
+@lru_cache
```

Comment:
Hm, I'm mostly indifferent on whether or not we keep this, but could this be premature optimization? I feel like it's pretty unlikely that the computation time to normalize the relatively short strings that we'll tend to find as header names will make much of a difference.

Reply:
Yeah, I used code from Tornado utils for that, just to be independent in case we will move to another framework https://github.com/tornadoweb/tornado/blob/58a548272d17cac8895c269be8f2d5b5cf62e5e3/tornado/httputil.py#L76

I agree that the optimization is very minor; the idea is that most of the time, it would normalize absolutely the same names (common header names) so we can cache it.